onnxruntime/core/providers/cuda/math/topk_impl.cu

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "topk_impl.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/cuda_common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void TopKKernel(const T* input_x, T* output_v, int64_t* output_i, const int64_t* elem_nums, size_t size, int64_t axis, int64_t K, int64_t largest, int64_t sorted, int64_t N, int64_t dimension) {
```

Comment:
>TopKKernel [](start = 16, length = 10)

Is this faster than radix sort from CUB? [https://github.com/dmlc/cub/blob/master/cub/device/device_radix_sort.cuh](https://github.com/dmlc/cub/blob/master/cub/device/device_radix_sort.cuh)

Reply:
Those may help too:
[https://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf](https://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf)
[https://github.com/anilshanbhag/gpu-topk](https://github.com/anilshanbhag/gpu-topk) 
[https://github.com/tensorflow/tensorflow/issues/5719](https://github.com/tensorflow/tensorflow/issues/5719)

---
In reply to: [328254646](https://github.com/microsoft/onnxruntime/pull/1919#discussion_r328254646) [](ancestors = 328254646)

Reply:
radix selection algorithm seems to be a good choice. I saw another one besides gpu-topk: https://github.com/facebook/fbcuda/blob/master/TopK.cuh

Reply:
It seems like both PyTorch and TF are using CUB, Cub is required for NomMaxSuppression op as well, we might want to include it as dependency.

Reply:
Using Cub now.